test/Table_Tests/src/Database/Upload_Spec.enso

```diff
@@ -837,6 +838,35 @@ spec make_new_connection prefix persistent_connector=True =
                 err.expected_type.is_text . should_be_true
                 err.got_type.is_numeric . should_be_true
 
+        dest_name = Name_Generator.random_name "dest-table-test-types"
+        structure =
+            x = Column_Description.Value "X" (Value_Type.Integer Bits.Bits_16)
+            y = Column_Description.Value "Y" (Value_Type.Char size=4 variable_length=True)
+            z = Column_Description.Value "Z" (Value_Type.Char size=5 variable_length=False)
+            [x, y, z]
+        dest = connection.create_table dest_name structure temporary=True primary_key=[]
+        non_trivial_types_supported =
+            has_warning_or_error = dest.is_error || (Problems.get_attached_warnings dest . not_empty)
+            has_warning_or_error.not
+        if non_trivial_types_supported then
+            src = source_table_builder [["X", [1, 2, 3]], ["Y", ["a", "xyz", "abcdefghijkl"]], ["Z", ["a", "pqrst", "abcdefghijkl"]]]
+            ## TODO [RW] figure out what semantics we want here; I think the current one may be OK but it is going to
+               be slightly painful, so IMO an auto-conversion could be useful. We could make it so that we do
+               auto-conversion (cast), but in a more strict mode such that if anything does not fit
+               (even just string padding required) we fail hard and tell the user to fix this.
+            Test.specify "fails if the target type is more restrictive than source" <|
+                result = src.update_database_table dest update_action=Update_Action.Insert key_columns=[]
+                result.should_fail_with Column_Type_Mismatch
+            but_maybe="I'm not sure if we want this automatic restriction. If anything, we should probably report situations like abcdefghijkl being truncated."
+            Test.specify "should warn if the target type is more restrictive than source and truncation may occur" pending=but_maybe <|
+                result = src.update_database_table dest update_action=Update_Action.Insert key_columns=[]
+                IO.println (Problems.get_attached_warnings result)
+
+                result.column_names . should_equal ["X", "Y", "Z"]
+                result.at "X" . to_vector . should_contain_the_same_elements_as [1, 2, 3]
+                result.at "Y" . to_vector . should_contain_the_same_elements_as ["a", "xyz", "abc"]
+                result.at "Z" . to_vector . should_contain_the_same_elements_as ["a     ", "pqrst", "abcde"]
```

Comment:
So this is not really for this PR, hence I'm not going to stop that PR by it.

But I just realised that if we have a table with e.g. 16-bit integers, and we create a table in Enso it will be 64-bit by default. So as shown in the test above, it will raise an error and require the user to `cast`.

I'm wondering if it would not be better to try casting automatically. If the cast has any warnings - error hard and get the user to resolve the situation, but if the types fit I imagine we could do this automatically for convenience.

Although that will only work in-memory where we can easily check if cast worked without warnings. In database we'd risk losing data so we surely cannot do this automatic conversion.

@jdunkerley do you think such convenience auto-cast in upload is worth it? If so, I will appreciate creating a ticket for it. Or just let me know and I'll create one once I'm back from vacation.

Reply:
I think this is worth doing; if the user has a target table with a narrow type, they are likely trying to write data that they believe will fit, so this is a common case.